### DISCUSSION TITLE
How to quickly delete the entire method?

### DISCUSSION THREAD

**Discussion by yaoqiangpersonal** [JetBrains/ideavim#374] [Q&A] [answered]
like this title, I want to find a way that can quickly delete the entire method. But I didn't find it in the end. Can anyone tell me?

**Comment by lippfi** [comment, marked as answer]
You can create some clever vim mapping to do this.
e.g. if you place your caret at the first line of function definition `nnoremap dm Va{%d` would do the job for functions like this:
```
fun name(args): Int {
  ....
}
```
or even better `nnoremap dk V/{<CR>%d` would also work for functions where arguments occupy multiple lines:
```
fun name(arg1,
arg2) {
  ...
}
```

**Reply by yaoqiangpersonal** [reply]
thanks a million. I use the way that it is like you supply.

**Reply by Sebastian-Nielsen** [reply]
@magidc's solution is better as it allows you to delete the method when executing the command from wherever inside the target method.
This solution only works if the closest enclosed curly-brackets surrounding the caret are the curly brackets of the function.

**Reply by lippfi** [reply]
@Sebastian-Nielsen you are right
by the way the issue is related to https://youtrack.jetbrains.com/issue/VIM-480/Add-text-objects-for-code-constructs-like-a-method-and-a-class . You can upvote the ticket or follow it's progress if you want (the more upvotes ticket has the more chances it will be implemented soon)

**Comment by magidc** [comment]
This is my solution:
``nnoremap dam :action MethodUp<cr>v0Of{%d``
It is based on a remap but it uses and action from Intellij. With this approach you can be in any position inside a method and still will do the job.

**Comment by magidc** [comment]
I have added a small modification to the mappings that allows to copy, delete or change methods (whole or just body) from any position of the method body or declaration:
```
nnoremap yim j:action MethodUp<cr>f{vi{y
nnoremap dim j:action MethodUp<cr>f{vi{d
nnoremap cim j:action MethodUp<cr>f{vi{c
nnoremap yam j:action MethodUp<cr>v0Of{%y
nnoremap dam j:action MethodUp<cr>v0Of{%d
nnoremap cam j:action MethodUp<cr>v0Of{%c

```
You can find my complete .ideavim config file [here](https://github.com/magidc/dotfiles/blob/master/.ideavimrc)

**Comment by victorivanov** [comment]
Wow, somehow I missed that we have access to Intellij commands in ex, the, at is a game changer for me! Thanks for sharing the solutions above. I wrote my own variations, which are more general but with the trade-off of `i` ones not working with multi-line method declarations (I have a feeling there is a solution for that somewhere though...). 

Those don't rely on curly braces so work for more languages and contexts, instead using the vim-standard `]m` motion). Note that I switched `m` to `f` as that's what I'm used to from the nvim plugin I'm used to. 

```
nnoremap Vaf :action MethodUp<cr>V]M
nnoremap vaf :action MethodUp<cr>V]M

nnoremap Vif :action MethodUp<cr>jV]Mk
nnoremap vif :action MethodUp<cr>jV]Mk

nnoremap daf :action MethodUp<cr>V]Md
nnoremap dif :action MethodUp<cr>jV]Mkd

nnoremap yaf :action MethodUp<cr>V]My
nnoremap yif :action MethodUp<cr>jV]Mky

nnoremap caf :action MethodUp<cr>V]Mc
nnoremap cif :action MethodUp<cr>jV]Mkc

```

Until today I was using the indent text objects provided by michaeljsmith/vim-indent-object (still a good workaround for those `i`/multi-line signatures)

**Reply by magidc** [reply]
Nice solution. I was not totally satisfied with the fact of using {} to match the limits of the method. Thx!

**Reply by megawubs** [reply]
I found that this solution also selected the opening bracket for a method body when its on a new line. I've updated it to this:

```
" Remaps to work with methods as text-objects
nnoremap Vam :action MethodUp<cr>V]M
nnoremap vam :action MethodUp<cr>V]M

nnoremap Vim :action MethodUp<cr>jjV]Mk
nnoremap vim :action MethodUp<cr>jjV]Mk

nnoremap dam :action MethodUp<cr>V]Md
nnoremap dim :action MethodUp<cr>jjV]Mkd

nnoremap yam :action MethodUp<cr>V]My
nnoremap yim :action MethodUp<cr>jjV]Mky

nnoremap cam :action MethodUp<cr>V]Mc
nnoremap cim :action MethodUp<cr>jjV]Mkc
```

**Reply by magidc** [reply]
As IdeaVIM plugin can be used in other Jetbrains IDEs besides Intellij (Java), I am trying to make it work for Pycharm (Python) but somehow text-objects seem to not work well. Anyone has tried to make a mapping compatible for Python?

**Reply by megawubs** [reply]
My experience thus far is not that good with the above mapping. For example, when i'm at the line of a method declaration, it goes one method up and deletes that one. It also does not delete the doc blocks when they are present. 

Ideally I would like a way to just describe text-objects based on language semantics as is possible with [nvim-tree-sitter-textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects).

**Reply by magidc** [reply]
Hi, my mappings cover the situation you has just described, when you are at the line of a method declaration. That's why there is "j" before every MethodUp action, basically it makes sure you are always under the desclaration so the MethodUp action moves to the declaration of the current method.
I could not solve the doc blocks issue though...

```vim
nnoremap vam j:action MethodUp<cr>0v]M
nnoremap vim j:action MethodUp<cr>]m<cr>v]M<bs>
nnoremap dam j:action MethodUp<cr>0v]M"_d
nnoremap dim j:action MethodUp<cr>]m<cr>v]M<bs>"_d
nnoremap yam j:action MethodUp<cr>0v]My
nnoremap yim j:action MethodUp<cr>]m<cr>v]M<bs>y
nnoremap cam j:action MethodUp<cr>0v]M"_c
nnoremap cim j:action MethodUp<cr>]m<cr>v]M<bs>"_c
```

**Reply by lippfi** [reply]
@megawubs > Ideally I would like a way to just describe text-objects based on language semantics as is possible with [nvim-tree-sitter-textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects)
It's doable (utilizing IJ Platform PSI); however, IdeaVim team is focused on other tasks, and there are no resources to implement and maintain this functionality. It sounds like a good idea for an IDE plugin that could be downloaded from marketplaces and used along with IdeaVim.
Implementing this in a separate plugin may still require changes to IdeaVim, and your PRs are always welcome.

**Reply by megawubs** [reply]
I do want to look into writing a plugin. But java/kotlin are not my primary languages and it feels a bit daunting to start something big like this.

**Reply by magidc** [reply]
@megawubs I feel the same. It would be awesome to have a treesitter like plugin to manage all kinds of text objects but I don't feel in mood to accomplish such a complex development. By the moment, I am staying with keymapping hacks to deal with it

**Reply by magidc** [reply]
I have just realized that there is a plugin that enables a text object for methods: [FunctionTextObj](https://plugins.jetbrains.com/plugin/25897-vim-functiontextobj)
Finally somebody was able to find the motivation to develop a plugin to solve this issue.
Contrary to our previously discussed mapping solutions, the plugin is able to select methods in several languages. It also selects method annotations

**Reply by megawubs** [reply]
Yes! This is what I was looking for. Thanks for mentioning it! This should now be the accepted answer.

**Reply by magidc** [reply]
Hey, I am doing a little promotion of my own plugins here. I have released a plugin that provides a very useful set of text objects including *classes*, *methods/functions*, *loops*, *block comments* or *arguments* among others. It is already available on Jetbrains marketplace: [vim-anyobject](https://plugins.jetbrains.com/plugin/28333-vim-anyobject).

It is a all in one plugin that gives you awesome productivity. It also solve some issues from other IdeaVim plugins like argtextobj (i.e. it does not work well for parametrized type arguments). 

Also, take a look to my other IdeaVim plugin [vim-dial](https://plugins.jetbrains.com/plugin/28237-vim-dial). It will enhance your productivity with the ability of incrementing/decrementing/switching between values without having to write it from the scratch (i.e. private-protected-public)

Thanks!!

**Reply by megawubs** [reply]
That looks amazing! I'm going to try this out.

**Reply by Sebastian-Nielsen** [reply]
Thank you for sharing magidc! Is it hard to develop plugins in ideavim? I only got experience with developing extensions in neovim. I pressume you have less freedom in ideavim plugins compared to neovim plugins, right?

**Reply by magidc** [reply]
it is not actually that hard once you understand the Intellij model. There were parts where I struggled because lack of documentation but once I managed to solve those situations all was quite straight forward. Now that the base is done, it is very easy to extend and define new features. 

If you miss some text object in AnyObject or some new swichable tags in Dial, feel free to ask, I would add them with pleasure

**Reply by huyz** [reply]
@magidc oh cool. Does it completely replace the functionalities of these plugins?
```vim
Plug 'vim-scripts/argtextobj.vim'
Plug 'kana/vim-textobj-entire'
```

I assume this one isn't covered at all:
```vim
Plug 'michaeljsmith/vim-indent-object'
```

**Reply by OptimusCrime** [reply]
I replaced [Vim Switch](https://plugins.jetbrains.com/plugin/25899-vim-switch) and [Vim FunctionTextObj](https://plugins.jetbrains.com/plugin/25897-vim-functiontextobj) with @magidc 's two plugins, as they offer more functionality.

Great work by the way!

**Reply by magidc** [reply]
> I replaced [Vim Switch](https://plugins.jetbrains.com/plugin/25899-vim-switch) and [Vim FunctionTextObj](https://plugins.jetbrains.com/plugin/25897-vim-functiontextobj) with @magidc 's two plugins, as they offer more functionality.
> 
> Great work by the way!

Thanks!!! Many more features to come :)

**Reply by magidc** [reply]
> @magidc oh cool. Does it completely replace the functionalities of these plugins?
> 
> ```viml
> Plug 'vim-scripts/argtextobj.vim'
> Plug 'kana/vim-textobj-entire'
> ```
> 
> I assume this one isn't covered at all:
> 
> ```viml
> Plug 'michaeljsmith/vim-indent-object'
> ```
Thanks you!!

Indent object will be included in the next release (this week). Condiftionals (if,switch...) will be also included.
If you find more case you would like support or you have any feedback, feel free to comment

**Comment by magidc** [comment]
AnyObject version 1.3.0 released: https://plugins.jetbrains.com/plugin/28333-vim-anyobject
It includes text objects for conditionals (if-else, switch, try-catch...) and for indented text blocks (like vim-indent-object)
